### PR TITLE
[WIP] SALTO-789: Use Salesforce deploy endpoint to deploy metadata instances

### PR DESCRIPTION
**Pull request by ori-moisis**
- Use the deploy endpoint in the metadata api to deploy all metadata instances in a single API call
- Currently this deploy allows partial success to maintain the behavior of the adapter as it was
- Changed preDeploy and onDeploy filter hooks to allow a filter to keep context between them and to allow filters to modify the change list like we do in onFetch

---

Still missing from this PR:
- Unit tests for workflow filter
- screenshots of the UX changes

_Release Notes_
- Use the deploy endpoint in the metadata api to deploy all metadata instances in a single API call (no rollback on error)

### CODE REVIEW

File: packages/salesforce-adapter/src/adapter.ts [hunk 1]
```diff
@@ -333,18 +308,45 @@ export default class SalesforceAdapter implements AdapterOperations {
...
-    metadataTypesToSkipMutation = [
-      'Workflow', // handled in workflow filter
-    ],
-    metadataTypesToUseUpsertUponUpdate = [
-      'Flow', // update fails for Active flows
-      'EscalationRules',
-      'AutoResponseRules',
-      'MatchingRules',
-    ],
+    // TODO:ORI - these types where using upsert instead of update, need to check they still work:
```
Comment (adamro) [inline packages/salesforce-adapter/src/adapter.ts:311]: reminder :)

File: packages/salesforce-adapter/src/adapter.ts [hunk 2]
```diff
@@ -492,33 +491,28 @@ export default class SalesforceAdapter implements AdapterOperations {
...
-            )
+        mainElemChanges.map(
+          objectChange => this.deployCustomObjectChanges(
+            topLevelNameToChanges[getChangeElement(objectChange).elemID.getFullName()],
+            objectChange,
+          )
         )
       )
+    } else {
+      // Deploy normal metadata instances
```
Comment (adamro) [inline packages/salesforce-adapter/src/adapter.ts:514]: define "normal" :) 
not important but I think this comment is redundant tbh

Comment (adamro) [inline packages/salesforce-adapter/src/adapter.ts:514]: It shouldn't happen and it's not a new case, but might we need to defend ourselves from a heterogeneous changeGroup that has changes of diff types (ie. CustomObjectInstance changes with CustomObject changes)

File: packages/salesforce-adapter/src/adapter.ts [hunk 3]
```diff
@@ -791,33 +650,27 @@ export default class SalesforceAdapter implements AdapterOperations {
...
+        await this.removeCustomObject(mainChange.data.before)
         return {
           appliedChanges: [mainChange],
           errors: [],
         }
       }
-      const after = await this.update(
-        mainChange.data.before, mainChange.data.after, elemChanges,
+      const after = await this.updateCustomObject(
+        mainChange.data.before, mainChange.data.after, elemChanges as Change<ObjectType | Field>[],
```
Comment (adamro) [inline packages/salesforce-adapter/src/adapter.ts:673]: nit: Shouldn't this be `as ReadonlyArray<Change<ObjectType | Field>>`?

Also, if it needs to be casted here and we didn't do anything with it before, wouldn't it make sense that this func also gets `elemChanges` as `elemChanges: ReadonlyArray<Change<ObjectType | Field>>`? (and maybe add a check before calling `deployCustomObjectChanges` that the changes actually only include ObjectType/Field)

File: packages/salesforce-adapter/src/filter.ts [hunk 4]
```diff
@@ -36,8 +36,10 @@ export type FilterCreator = (
 export const filtersRunner = (client: SalesforceClient,
   config: FilterContext,
   filterCreators: ReadonlyArray<FilterCreator>): Required<Filter> => {
+  const allFilters = filterCreators.map(f => f({ client, config }))
```
Comment (adamro) [inline packages/salesforce-adapter/src/filter.ts:39]: nit: why? :)

File: packages/salesforce-adapter/src/group_changes.ts [hunk 5]
```diff
@@ -14,24 +14,50 @@
...
-  ): change is ChangeEntry<InstanceElement> => {
-    const changeElement = getChangeElement(change[1])
-    return isInstanceOfCustomObject(changeElement)
-  }
-  const instanceOfCustomObjectChangeToGroupId = (change: Change<InstanceElement>): string =>
-    `${change.action}_${apiName(getChangeElement(change).type)}_instances`
+type ChangeIdFunction = (change: Change) => string | undefined
 
-  return new Map(
+const isInstanceOfCustomObjectChange = (
```
Comment (adamro) [inline packages/salesforce-adapter/src/group_changes.ts:23]: We already have very similar code (exported from `custom_object_instances_deploy.ts` for change*s*)

**Comment by ori-moisis** [comment]
will re-open once custom object support is implemented

**Pull request closed without merging**